File: problems/1406_stone-game-iii/solution.py
from typing import List, Optional


class Solution:
    def stoneGameIII(self, stoneValue: List[int]) -> str:
        n = len(stoneValue)
        dp = [0] * (n + 1)
        for i in range(n - 1, -1, -1):
            take = 0
            best = float("-inf")
            for j in range(i, min(i + 3, n)):
                take += stoneValue[j]
                cand = take - dp[j + 1]
                if cand > best:
                    best = cand
            dp[i] = best
        if dp[0] > 0:
            return "Alice"
        if dp[0] < 0:
            return "Bob"
        return "Tie"

File: problems/1406_stone-game-iii/solution.cpp
class Solution {
  public:
    string stoneGameIII(vector<int> &stoneValue) {
        int n = stoneValue.size();
        vector<long long> dp(n + 1, 0);
        for (int i = n - 1; i >= 0; i--) {
            long long take = 0;
            long long best = LLONG_MIN;
            int hi = min(i + 3, n);
            for (int j = i; j < hi; j++) {
                take += stoneValue[j];
                long long cand = take - dp[j + 1];
                best = max(best, cand);
            }
            dp[i] = best;
        }
        if (dp[0] > 0) {
            return "Alice";
        }
        if (dp[0] < 0) {
            return "Bob";
        }
        return "Tie";
    }
};
